encoding/ccf/encode.go

```diff
@@ -0,0 +1,1900 @@
+		visited[cadenceTypeID] = ccfTypeID(len(visited))
+	}
+
+	simpleTypeID, ok := simpleTypeIDByType(typ)
+	if ok {
+		return e.encodeSimpleTypeValue(simpleTypeID)
+	}
+
+	switch typ := typ.(type) {
+	case *cadence.OptionalType:
+		return e.encodeOptionalTypeValue(typ, visited)
+
+	case *cadence.VariableSizedArrayType:
+		return e.encodeVarSizedArrayTypeValue(typ, visited)
+
+	case *cadence.ConstantSizedArrayType:
+		return e.encodeConstantSizedArrayTypeValue(typ, visited)
+
+	case *cadence.DictionaryType:
+		return e.encodeDictTypeValue(typ, visited)
+
+	case *cadence.StructType:
+		return e.encodeStructTypeValue(typ, visited)
+
+	case *cadence.ResourceType:
+		return e.encodeResourceTypeValue(typ, visited)
+
+	case *cadence.EventType:
+		return e.encodeEventTypeValue(typ, visited)
+
+	case *cadence.ContractType:
+		return e.encodeContractTypeValue(typ, visited)
+
+	case *cadence.StructInterfaceType:
+		return e.encodeStructInterfaceTypeValue(typ, visited)
+
+	case *cadence.ResourceInterfaceType:
+		return e.encodeResourceInterfaceTypeValue(typ, visited)
+
+	case *cadence.ContractInterfaceType:
+		return e.encodeContractInterfaceTypeValue(typ, visited)
+
+	case *cadence.FunctionType:
+		return e.encodeFunctionTypeValue(typ, visited)
+
+	case *cadence.ReferenceType:
+		return e.encodeReferenceTypeValue(typ, visited)
+
+	case *cadence.RestrictedType:
+		return e.encodeRestrictedTypeValue(typ, visited)
+
+	case *cadence.CapabilityType:
+		return e.encodeCapabilityTypeValue(typ, visited)
+
+	case *cadence.EnumType:
+		return e.encodeEnumTypeValue(typ, visited)
+
+	default:
+		panic(cadenceErrors.NewUnexpectedError("unsupported type value %s (%T)", typ.ID(), typ))
+	}
+}
+
+// encodeNullableTypeValue encodes cadence.Type as type-value or nil.
+func (e *Encoder) encodeNullableTypeValue(typ cadence.Type, visited ccfTypeIDByCadenceType) error {
+	if typ == nil {
+		return e.encodeNilTypeValue()
+	}
+	return e.encodeTypeValue(typ, visited)
+}
+
+// encodeTypeValueRef encodes type value ID as
+// language=CDDL
+// type-value-ref =
+//
+//	; cbor-tag-type-value-ref
+//	#6.184(id)
+func (e *Encoder) encodeTypeValueRef(id ccfTypeID) error {
+	rawTagNum := []byte{0xd8, CBORTagTypeValueRef}
+	return e.encodeTypeRefWithRawTag(id, rawTagNum)
+}
+
+// encodeSimpleTypeValue encodes cadence simple type value as
+// language=CDDL
+// simple-type-value =
+//
+//	; cbor-tag-simple-type-value
+//	#6.185(simple-type-id)
+func (e *Encoder) encodeSimpleTypeValue(id uint64) error {
+	rawTagNum := []byte{0xd8, CBORTagSimpleTypeValue}
+	return e.encodeSimpleTypeWithRawTag(id, rawTagNum)
+}
+
+// encodeOptionalTypeValue encodes cadence.OptionalType as
+// language=CDDL
+// optional-type-value =
+//
+//	; cbor-tag-optional-type-value
+//	#6.186(type-value)
+func (e *Encoder) encodeOptionalTypeValue(typ *cadence.OptionalType, visited ccfTypeIDByCadenceType) error {
+	rawTagNum := []byte{0xd8, CBORTagOptionalTypeValue}
+	return e.encodeOptionalTypeWithRawTag(
```

Comment:
Nice work abstracting the encoding and decoding of the two different type hierarchies 👏